### PR TITLE
Import/Export - Fix 'entity values are mixed' exception when no valid data

### LINKED ISSUE #109: Error in data structure: entity values are mixed
hello, have this error when try to import with sample code, Error in data structure: entity values are mixed

why??

```



use Magento\Framework\App\Bootstrap;
require_once(__DIR__.'/../../app/bootstrap.php');
$bootstrap = Bootstrap::create(BP, $_SERVER);
$objectManager = \Magento\Framework\App\ObjectManager::getInstance(); 
$obj = $bootstrap->getObjectManager();

$state = $obj->get('Magento\Framework\App\State');
$state->setAreaCode('frontend');

$importerModel = $objectManager->create('FireGento\FastSimpleImport\Model\Importer');

$productsArray[] = array(
        
                    'sku' => 'firegento-test',
                    'attribute_set_code' => 'Default',
                    'product_type' => 'simple',
                    'product_websites' => 'base',
                    'name' => 'FireGento Test Product',
                    'price' => '14',
                );
                
                
try {
    $importerModel->processImport($productsArray);
} catch (\Exception $e) {
    echo $e->getMessage();
}

```

### LINKED ISSUE #32905: [Issue] Import/Export - Fix 'entity values are mixed' exception when no valid data
This issue is automatically created based on existing pull request: magento/magento2#32904: Import/Export - Fix 'entity values are mixed' exception when no valid data

---------
### Description
The Import/Export module can operate in two modes: skip errors, stop on error. When configured to skip errors, the import will proceed irrespective of the validation result. When there is no data to import, a exception "entity values are mixed" is thrown. This pull request solves the problem by failing validation in this case.

### Fixed Issues
1. Fixes https://github.com/firegento/FireGento_FastSimpleImport2/issues/109

### Manual testing scenarios
1. Create an import file with only invalid data (eg, SKUs that don't exist, invalid customer address fields, malformed customer email addresses)
1. Configure the Magento importer to skip validation errors
1. Import the file
1. Without this pull request: validation passes, but the import fails with a message that "entity values are mixed"
1. With this pull request: validation fails with a message saying that "There are no valid rows to import."

### Questions or comments
*none*

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by fredden**
### Description
The Import/Export module can operate in two modes: skip errors, stop on error. When configured to skip errors, the import will proceed irrespective of the validation result. When there is no data to import, a exception "entity values are mixed" is thrown. This pull request solves the problem by failing validation in this case.

### Fixed Issues
1. Fixes https://github.com/firegento/FireGento_FastSimpleImport2/issues/109 
1. Fixes magento/magento2#32905

### Manual testing scenarios
1. Create an import file with only invalid data (eg, SKUs that don't exist, invalid customer address fields, malformed customer email addresses)
1. Configure the Magento importer to skip validation errors
1. Import the file
1. Without this pull request: validation passes, but the import fails with a message that "entity values are mixed"
1. With this pull request: validation fails with a message saying that "There are no valid rows to import."

### Questions or comments
*none*

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [x] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#32905: Import/Export - Fix 'entity values are mixed' exception when no valid data

### CODE REVIEW

**Comment by fredden** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento create issue

**Comment by fredden** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please check failing integration tests

**Review by Den4ik — CHANGES_REQUESTED** [review]
Hi @fredden! Thanks for your contribution.
I believe that extending of `isErrorLimitExceeded` for both situations: `skip` or `stop` on errors is better solution.

**Comment by fredden** [comment]
> I believe that extending of `isErrorLimitExceeded` for both situations: `skip` or `stop` on errors is better solution.

@Den4ik the `ProcessingErrorAggregator` (where `isErrorLimitExceeded()` is found) does not appear to have access to the total row count. How do you suggest this information should be obtained?

**Comment by fredden** [comment]
@magento run all tests

File: app/code/Magento/ImportExport/Model/Import.php [hunk 1]
```diff
@@ -683,9 +683,16 @@ public function validateSource(AbstractSource $source)
         $this->addLogComment($messages);
 
         $result = !$errorAggregator->isErrorLimitExceeded();
+
+        if ($this->getProcessedRowsCount() <= $errorAggregator->getInvalidRowsCount()) {
+            $result = false;
+            $this->addLogComment(__('There are no valid rows to import.'));
```
Comment (Den4ik) [inline app/code/Magento/ImportExport/Model/Import.php]:
```suggestion
            $this->addLogComment(__('There are no valid rows to import.'));
            return false;
```

**Review by Den4ik — CHANGES_REQUESTED** [review]
> @Den4ik the ProcessingErrorAggregator (where isErrorLimitExceeded() is found) does not appear to have access to the total row count. How do you suggest this information should be obtained?

@fredden You were right and your solution look clear. Just think that direct return false result allow skip `if` check on next step

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Comment by fredden** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B
Looks good to me

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]
✔ Approved.

Failing tests looks not related to changes from this PR.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9084](https://jira.corp.magento.com/browse/ENGCOM-9084) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Review by Den4ik — APPROVED** [review]
✅ Approved.

Agree with @ihor-sviziev, failed test looks like not related. But one more time

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-9084](https://jira.corp.magento.com/browse/ENGCOM-9084) has been created to process this Pull Request

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Comment by Den4ik** [comment]
@magento run Functional Tests B2B

**Comment by fredden** [comment]
@magento run all tests

**Comment by fredden** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Lima** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Lima** [comment]
@magento run all tests

**Review by engcom-Lima — COMMENTED** [review]
Hi @fredden , 

Thanks for the collaboration & contribution!

### :x: QA not Passed

Still, I am getting the same error before and after the PR changes "_Additional data: <div class="messages"><div class=" message message-error error"><div data-ui-id="messages-message-error" >Error in data structure: entity values are mixed</div></div></div>_"

**Manual Scenario Steps**
1. Create an import file with only invalid data (eg, SKUs that don't exist, invalid customer address fields, malformed customer email addresses)
2. Configure the Magento importer to skip validation errors
3. Import the file

**:heavy_check_mark: Expected result After Fix** _Validation fails with a message saying that "There are no valid rows to import."_

**:x: Actual result After Fix** 
Screenshot 2023-06-23 at 1 17 21 PM

Kindly recheck it. If we are missing anything please inform us. 

Thanks

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
### :heavy_check_mark: QA Passed

**Preconditions:**
- Install fresh Magento **`2.4-develop`** and **`PHP 8.1`**

**Manual testing scenario:*

1. Create an import file with only invalid data (eg, SKUs that don't exist, invalid customer address fields, malformed customer email addresses)
2. Configure the Magento importer to skip validation errors
3. Import the file

**Before:** :heavy_multiplication_x: Import button was displayed.

Screenshot 2023-07-06 at 12 40 30 PM

**Getting error after importing "_Additional data:Error in data structure: entity values are mixed_":**
Screenshot 2023-06-23 at 4 31 02 PM

**After:** :heavy_check_mark:  Now working fine.
_**Import button disappeared because of invalid data**_
Screenshot 2023-07-06 at 12 40 48 PM

Screenshot 2023-07-06 at 12 58 16 PM
Screenshot 2023-07-06 at 12 57 06 PM


Builds are failed. Hence, moving this PR to **Extended Testing.**

**Comment by engcom-Echo** [comment]
@magento run Integration Tests, Functional Tests EE,Functional Tests CE,Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Integration Tests

**Comment by engcom-Echo** [comment]
Functional Tests CE, Functional Tests EE and Functional Tests B2B failures are different on last two run on same commit.
Functional Tests CE
Run1
image

Run2
image


Functional Tests EE
Run1
image

Run2
image

Functional Tests B2B
Run1
image

Run2
image

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
**Integration Tests** failure is not related to PR.
**Functional Tests CE, Functional Tests EE, Functional Tests B2B** failures are different on last two run on same commit.
**Functional Tests CE**
Run1
image

Run2
image

**Functional Tests EE**
Run1
image

Run2
image

**Functional Tests B2B**
Run1
image

Run2
image

Hence moving it to `Merge In Progress`

**Pull request merged**